### PR TITLE
Fixing the warm-up issue 542

### PR DESCRIPTION
**Pull request by Jd8111997**
Issue #542 

This is the command that I used to run linter on the primary files that I modified : 
```bash
i lint --files "dataflow/core/utils.py dataflow/core/test/test_utils.py dev_scripts/cleanup_scripts/SorrTask542_convert_to_multiindex.sh"
```
It's is only failing for amp_pylint on dataflow/core/utils.py file and it's failing on the code that were preexisting before and I haven't modified anything in that file except making the existing function public.
This is the output of linter for failing case:
```bash
amp_processjupytext......................................................Passed
amp_check_filename.......................................................Passed
amp_warn_incorrectly_formatted_todo......................................Passed
amp_flake8...............................................................Passed
amp_pylint...............................................................Failed
- hook id: amp_pylint
- exit code: 5

One and only one set-up config should be true:
is_cmamp_prod=False
is_dev4=False
is_dev_ck=False
is_ig_prod=False
is_inside_ci=False
is_mac=False
INFO: > cmd='/app/linters/amp_pylint.py dataflow/core/utils.py dataflow/core/test/test_utils.py'
************* Module dataflow.core.utils
dataflow/core/utils.py:228: [R1704(redefined-argument-from-local), convert_to_multiindex] Redefining argument with the local name 'df'
dataflow/core/utils.py:277: [C0209(consider-using-f-string), convert_to_list] Formatting a regular string which could be a f-string
dataflow/core/utils.py:304: [C3001(unnecessary-lambda-assignment), get_forward_cols.<lambda>] Lambda expression assigned to a variable. Define a function using the "def" keyword instead.
dataflow/core/utils.py:304: [C0209(consider-using-f-string), get_forward_cols.<lambda>] Formatting a regular string which could be a f-string

amp_mypy.................................................................Failed
- hook id: amp_mypy
- exit code: 3

One and only one set-up config should be true:
is_cmamp_prod=False
is_dev4=False
is_dev_ck=False
is_ig_prod=False
is_inside_ci=False
is_mac=False
INFO: > cmd='/app/linters/amp_mypy.py dataflow/core/utils.py dataflow/core/test/test_utils.py'
dataflow/core/utils.py:223: error: Need type annotation for 'dfs' (hint: "dfs: Dict[<type>, <type>] = ...")  [var-annotated]
dataflow/core/utils.py:399: error: Value of type "Optional[Match[str]]" is not indexable  [index]
dataflow/core/utils.py:403: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]

1
00:12:15 - INFO  lib_tasks_lint.py lint:409                             Parsing 'linter_output.txt'
00:12:15 - INFO  lib_tasks_lint.py lint:414                             Saved cfile in './linter_warnings.txt'
dataflow/core/utils.py:223:[amp_mypy] error: Need type annotation for 'dfs' (hint: "dfs: Dict[<type>, <type>] = ...")  [var-annotated]
dataflow/core/utils.py:228:[amp_pylint] [R1704(redefined-argument-from-local), convert_to_multiindex] Redefining argument with the local name 'df'
dataflow/core/utils.py:277:[amp_pylint] [C0209(consider-using-f-string), convert_to_list] Formatting a regular string which could be a f-string
dataflow/core/utils.py:304:[amp_pylint] [C0209(consider-using-f-string), get_forward_cols.<lambda>] Formatting a regular string which could be a f-string
dataflow/core/utils.py:304:[amp_pylint] [C3001(unnecessary-lambda-assignment), get_forward_cols.<lambda>] Lambda expression assigned to a variable. Define a function using the "def" keyword instead.
dataflow/core/utils.py:399:[amp_mypy] error: Value of type "Optional[Match[str]]" is not indexable  [index]
dataflow/core/utils.py:403:[amp_mypy] error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
```

All the other changes are made by cleaning code base script for updating function calls and imports.

### CODE REVIEW

File: dataflow/core/test/test_utils.py [hunk 1]
```diff
@@ -48,3 +48,52 @@ def test3(self) -> None:
...
+            + ["2022-01-04 09:03:00-05:00"]
+        )
+        df = pd.DataFrame(data, index=index)
+        return df
+
+    def test1(self) -> None:
+        """
+        Test that the function transforms the DataFrame correctly.
+        """
+        df = self.get_test_data()
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:73]: Let's add comments to split the chunks of the code.
```
# Prepare inputs.
...
# Run the code.
...
# Check the output.
```

Can you propose a change to the doc to ./docs/coding/all.write_unit_tests.how_to_guide.md to explain this, if it's not there already?
You can do it in a different PR

File: dataflow/core/test/test_utils.py [hunk 2]
```diff
@@ -48,3 +48,52 @@ def test3(self) -> None:
...
+            ["2022-01-04 09:01:00-05:00"] * 2
+            + ["2022-01-04 09:02:00-05:00"] * 2
+            + ["2022-01-04 09:03:00-05:00"]
+        )
+        df = pd.DataFrame(data, index=index)
+        return df
+
+    def test1(self) -> None:
+        """
+        Test that the function transforms the DataFrame correctly.
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:71]: the -> a

File: dataflow/core/test/test_utils.py [hunk 3]
```diff
@@ -48,3 +48,52 @@ def test3(self) -> None:
...
+        df = self.get_test_data()
+        asset_id_col = "id"
+        result_df = dtfcorutil.convert_to_multiindex(df, asset_id_col)
+
+        expected_columns = pd.MultiIndex.from_product(
+            [["close", "volume"], [13684, 17085]], names=[None, "id"]
+        )
+        self.assert_equal(
+            str(expected_columns.to_list()), str(result_df.columns.to_list())
+        )
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:82]: We don't like copy/paste, not even in test code.
This can be factored out in an helper function.
You can ask @samarth9008 @sonaalKant how to do it.

Again feel free to improve the doc to clarify this

Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:82]: I'd add also a check_string to test even better.

File: dataflow/core/test/test_utils.py [hunk 4]
```diff
@@ -48,3 +48,52 @@ def test3(self) -> None:
...
+        asset_id_col = "id"
+
+        df = pd.concat([df, df.head(1)])
+        result_df = dtfcorutil.convert_to_multiindex(df, asset_id_col)
+        expected_columns = pd.MultiIndex.from_product(
+            [["close", "volume"], [13684, 17085]], names=[None, "id"]
+        )
+        self.assert_equal(
+            str(expected_columns.to_list()), str(result_df.columns.to_list())
+        )
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:99]: Other tests for the corner cases should be a df with only one asset.

Think about what other corner cases there can be. You can/should look at the function.

What happens with an empty df?
What if asset_id_col doesn't exist?
What if there is no index?

Your goal is to find corner cases where the function doesn't behave properly. Otherwise you just test what you know already know that works, which is not great.

Again pls improve the doc with this wisdom pills.

**Review by gpsaggese — COMMENTED** [review]
Not bad as first PR. Keep pushing.

**Comment by Jd8111997** [comment]
Thanks for the suggestions @gpsaggese , I will update the PR.

File: dataflow/core/test/test_utils.py (hunk 3 above)
Comment (Jd8111997) [reply dataflow/core/test/test_utils.py:82]: @samarth9008 @sonaalKant Can you please elaborate on this ?

Comment (Jd8111997) [reply dataflow/core/test/test_utils.py:82]: @samarth9008 @sonaalKant Can you please let me know about this ?

Comment (sonaalKant) [reply dataflow/core/test/test_utils.py:82]: Just move the repetitive code to a helper function. Eg : `_test_convert_to_multiindex`  is a new method which takes input df, expected df and assert on mismatch.

**Comment by Jd8111997** [comment]
@sonaalKant @samarth9008, I have updated the PR incorporating all suggestions provided by  @gpsaggese

**Comment by samarth9008** [comment]
Resolve the conversation if they are addressed.

File: research_amp/cc/notebooks/CMTask1704_Get_raw_CC_data.py [hunk 5]
```diff
@@ -28,6 +28,7 @@
 import core.finance.resampling as cfinresa
 import core.plotting.normality as cplonorm
 import dataflow.core as dtfcore
+import dataflow.core.utils as dtfcorutil
```
Comment (samarth9008) [inline research_amp/cc/notebooks/CMTask1704_Get_raw_CC_data.py:31]: No need of full import as we are outside the dataflow dir(package). The import in the above line will work. 

The idea is
When we are outside a package, e.g., `dataflow/core`, we import the whole package like we did here `import dataflow.core as dtfcore`. We do it even if we need to address only one function. The exception is for `helpers`, there're too much dependencies that can affect on a flow so we import it as modules even if we outside the package, e.g. `helpers.hgit`.

The opposite, if we would be inside the package, e.g., inside `dataflow`, we would import `import dataflow.core.utils as dtfcorutil` to prevent cycle imports that could ruin the code sometimes.

Simply saying, module is a py-file, package is a dir that contains py-files. Does it make sense?

Make changes everywhere

**Review by samarth9008 — CHANGES_REQUESTED** [review]
TODO(Samarth): Don't merge before creating a PR on cmamp and testing it over there also

File: research_amp/cc/notebooks/CMTask1704_Get_raw_CC_data.py (hunk 5 above)
Comment (Jd8111997) [reply research_amp/cc/notebooks/CMTask1704_Get_raw_CC_data.py]: Thanks @samarth9008 for the feedback. I updated the PR incorporating all your suggestion.

**Comment by gpsaggese** [comment]
Going in the right direction.
As discussed, you need to resolve all the comments and ask for a new review.

**Comment by Jd8111997** [comment]
@sonaalKant @samarth9008 I resolved and addressed all feedback in my latest PR. Could you please review it again ?

File: dataflow/core/test/test_utils.py [hunk 6]
```diff
@@ -48,3 +49,147 @@ def test3(self) -> None:
...
+        asset_id_col = "id"
+        df = pd.concat([df, df.head(1)])
+
+        # Prepare the actual output by running function.
+        actual_df = dtfcorutil.convert_to_multiindex(df, asset_id_col)
+    
+        # Compare the result.
+        self.assert_multiindex_columns_equal(expected_df_columns, actual_df)
+
+    def test3(self) -> None:
```
Comment (sonaalKant) [inline dataflow/core/test/test_utils.py:134]: Why -> None? when you are return boolean

File: dataflow/core/test/test_utils.py [hunk 7]
```diff
@@ -48,3 +49,147 @@ def test3(self) -> None:
...
+        Test that a function handles an empty dataframe correctly.
+        """ 
+        # Creating empty data frame for input.
+        df = pd.DataFrame(columns=['id', 'close', 'volume'])
+        asset_id_col = 'id'
+
+        # Method is successfully passing test case if it's throwing Error on empty df. 
+        try:
+            actual_df = dtfcorutil.convert_to_multiindex(df, asset_id_col)
+        except AssertionError:
```
Comment (sonaalKant) [inline dataflow/core/test/test_utils.py:145]: Instead of catching the Assertion error and returning True, get the error message and and assert_equal on expected error message. Otherwise this test case will pass if it fails on something else.

File: dataflow/core/test/test_utils.py [hunk 8]
```diff
@@ -48,3 +49,147 @@ def test3(self) -> None:
...
+        }
+        index = pd.to_datetime(['2022-01-04 09:01:00-05:00', '2022-01-04 09:02:00-05:00', '2022-01-04 09:03:00-05:00'])
+        df = pd.DataFrame(data, index = index)
+        expected_df_columns = pd.MultiIndex.from_product([['close', 'volume'], [13684]], names=[None, 'id'])
+        return df, expected_df_columns
+    
+    def assert_multiindex_columns_equal(self,
+                                        expected_df_columns: pd.MultiIndex,
+                                        actual_df: pd.DataFrame
+                                        ):
```
Comment (sonaalKant) [inline dataflow/core/test/test_utils.py:95]: ->
```
def assert_multiindex_columns_equal(
 self,
 expected_df_columns: pd.MultiIndex,
 actual_df: pd.DataFrame,
 ) -> None:
```

**Review by sonaalKant — CHANGES_REQUESTED** [review]
Almost there.

File: dataflow/core/test/test_utils.py [hunk 9]
```diff
@@ -48,3 +49,137 @@ def test3(self) -> None:
...
+            'close': [None, None, None],
+            'volume': [0, 0, 0]
+        }
+        index = pd.to_datetime(['2022-01-04 09:01:00-05:00', '2022-01-04 09:02:00-05:00', '2022-01-04 09:03:00-05:00'])
+        df = pd.DataFrame(data, index = index)
+        expected_df_columns = pd.MultiIndex.from_product([['close', 'volume'], [13684]], names=[None, 'id'])
+        return df, expected_df_columns
+    
+    def assert_multiindex_columns_equal(self,
+                                        expected_df_columns: pd.MultiIndex,
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:92]: Linter

File: dataflow/core/test/test_utils.py [hunk 10]
```diff
@@ -48,3 +49,137 @@ def test3(self) -> None:
...
+                                        actual_df: pd.DataFrame
+                                        ) -> None:
+        """
+        Compares the list of columns of actual df and and expected df.
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:96]: -> Compare

File: dataflow/core/test/test_utils.py [hunk 11]
```diff
@@ -48,3 +49,137 @@ def test3(self) -> None:
...
+        """
+        # Prepare input and expected output.
+        df = self.get_test_data_single_asset()
+        asset_id_col = 'nonexistent_column'
+        # Method is successfully passing test case if it's returning error message that is similar to expected error message.
+        with self.assertRaises(AssertionError) as context:
+            actual_df = dtfcorutil.convert_to_multiindex(df, asset_id_col)
+        # Get the actual error message
+        actual_error_message = str(context.exception)
+        expected_error_message = "\n################################################################################\n* Failed assertion *\nInstance of '(                              id close  volume\n2022-01-04 09:01:00-05:00  13684  None       0\n2022-01-04 09:02:00-05:00  13684  None       0\n2022-01-04 09:03:00-05:00  13684  None       0, MultiIndex([( 'close', 13684),\n            ('volume', 13684)],\n           names=[None, 'id']))' is '<class 'tuple'>' instead of '<class 'pandas.core.frame.DataFrame'>'\n################################################################################\n"
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:167]: I would do ''' so you don't have to use \n and it can be formatted better

File: dataflow/core/test/test_utils.py [hunk 12]
```diff
@@ -48,3 +49,137 @@ def test3(self) -> None:
...
+        """
+        # Prepare input and expected output.
+        df , expected_df_columns= self.get_test_data_single_asset()
+        asset_id_col = 'id'
+        df = df.reset_index(drop=True)
+        # Prepare the actual output by running function.
+        actual_df = dtfcorutil.convert_to_multiindex(df, asset_id_col)
+        # Compare the result.
+        self.assert_multiindex_columns_equal(expected_df_columns, actual_df)
+
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:182]: Linter

File: dataflow/core/utils.py [hunk 13]
```diff
@@ -181,6 +183,69 @@ def validate_df_indices(df: pd.DataFrame) -> None:
     # TODO(Paul): assert if the datetime index has dups.
 
 
+def convert_to_multiindex(df: pd.DataFrame, asset_id_col: str) -> pd.DataFrame:
+    """
+    Transform a df like: ```
```
Comment (gpsaggese) [inline dataflow/core/utils.py:188]: You want an empty line before ``` for the linter to be happy

**Review by gpsaggese — COMMENTED** [review]
Good work. Few nits and then merge.

Let's push on this so we can move to the next one

**Comment by Jd8111997** [comment]
Done, I also attached linter output for test_utils.py for your reference @gpsaggese.

**Review by sonaalKant — APPROVED** [review]
LGTM

File: linter_output.txt [hunk 14]
```diff
@@ -0,0 +1,197 @@
+The AM_AWS_ACCESS_KEY_ID variable is not set. Defaulting to a blank string.
```
Comment (gpsaggese) [inline linter_output.txt:1]: No need to add this to the PR. FYI the GH action automatically compute this

File: dev_scripts/client_setup/requirements.txt [hunk 15]
```diff
@@ -7,3 +7,4 @@ poetry
 pytest >= "6.0.0"
 s3fs  # For tools like `publish_notebook.py`.
 tqdm
+pyyaml == 5.3.1
```
Comment (gpsaggese) [inline dev_scripts/client_setup/requirements.txt:10]: Why is this needed?

File: dataflow/core/test/test_utils.py [hunk 16]
```diff
@@ -48,3 +49,140 @@ def test3(self) -> None:
         act = dtfcorutil.get_DagBuilder_name_from_string(dag_builder_ctor_as_str)
         exp = "C5b"
         self.assert_equal(act, exp)
+
+
+class Test_convert_to_multiindex(hunitest.TestCase):
+    @staticmethod
+    def get_test_data() -> Tuple[pd.DataFrame, pd.MultiIndex]:
+        """
+        Function that returns dummy dataframe with multiple asset id and
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:58]: Return ...

File: dataflow/core/test/test_utils.py [hunk 17]
```diff
@@ -48,3 +49,140 @@ def test3(self) -> None:
...
+        df = pd.DataFrame(data, index=index)
+        expected_df_columns = pd.MultiIndex.from_product(
+            [["close", "volume"], [13684, 17085]], names=[None, "id"]
+        )
+        return df, expected_df_columns
+
+    @staticmethod
+    def get_test_data_single_asset() -> Tuple[pd.DataFrame, pd.MultiIndex]:
+        """
+        " Function that returns dummy dataframe with single asset id and
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:80]: No "

-> Return ...

File: dataflow/core/test/test_utils.py [hunk 18]
```diff
@@ -48,3 +49,140 @@ def test3(self) -> None:
         act = dtfcorutil.get_DagBuilder_name_from_string(dag_builder_ctor_as_str)
         exp = "C5b"
         self.assert_equal(act, exp)
+
+
+class Test_convert_to_multiindex(hunitest.TestCase):
+    @staticmethod
+    def get_test_data() -> Tuple[pd.DataFrame, pd.MultiIndex]:
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:56]: -> get_test_data_multiple_assets for symmetry

File: dataflow/core/test/test_utils.py [hunk 19]
```diff
@@ -48,3 +49,140 @@ def test3(self) -> None:
...
+        """
+        Test that a function handles an empty dataframe correctly.
+        """
+        # Creating empty data frame for input.
+        df = pd.DataFrame(columns=["id", "close", "volume"])
+        asset_id_col = "id"
+        with self.assertRaises(AssertionError) as context:
+            dtfcorutil.convert_to_multiindex(df, asset_id_col)
+        actual_error_message = str(context.exception)
+        expected_error_message = '\n################################################################################\n* Failed assertion *\n1 <= 0\n################################################################################\n'
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:146]: You can use r""" to make it look better

File: dataflow/core/test/test_utils.py [hunk 20]
```diff
@@ -48,3 +49,140 @@ def test3(self) -> None:
...
+        exist.
+        """
+        # Prepare input and expected output.
+        df = self.get_test_data_single_asset()
+        asset_id_col = "nonexistent_column"
+        with self.assertRaises(AssertionError) as context:
+            dtfcorutil.convert_to_multiindex(df, asset_id_col)
+        # Get the actual error message
+        actual_error_message = str(context.exception)
+        expected_error_message = "\n################################################################################\n* Failed assertion *\nInstance of '(                              id close  volume\n2022-01-04 09:01:00-05:00  13684  None       0\n2022-01-04 09:02:00-05:00  13684  None       0\n2022-01-04 09:03:00-05:00  13684  None       0, MultiIndex([( 'close', 13684),\n            ('volume', 13684)],\n           names=[None, 'id']))' is '<class 'tuple'>' instead of '<class 'pandas.core.frame.DataFrame'>'\n################################################################################\n"
```
Comment (gpsaggese) [inline dataflow/core/test/test_utils.py:173]: Same r""" to make it readable

**Review by gpsaggese — COMMENTED** [review]
Final tweaks and merge for me

File: linter_output.txt (hunk 14 above)
Comment (Jd8111997) [reply linter_output.txt:1]: Alright!

File: dev_scripts/client_setup/requirements.txt (hunk 15 above)
Comment (Jd8111997) [reply dev_scripts/client_setup/requirements.txt:10]: Because I was getting `pyyaml` dependency error while I was trying to build virtual environment so I have to manually add this in `requirements.txt`. I mentioned it here: [Issue](https://github.com/sorrentum/sorrentum/issues/668#issuecomment-1962759845l)

File: dataflow/core/test/test_utils.py (hunk 19 above)
Comment (Jd8111997) [reply dataflow/core/test/test_utils.py:146]: I tried using multi-line string formatting, but to improve readability, I also need to add indentations. However, since the `actual_error_message` does not contain any indentations, the `assert_equal` method will fail. I attempted to remove the indentations from the multi-line string, but it ended up looking even more unappealing. I have attached an image for your reference."
oie_al2wAyCQsR1e

Comment (Jd8111997) [reply dataflow/core/test/test_utils.py:146]: Okay, thanks for suggestion @samarth9008, I will try and let you know.

Comment (Jd8111997) [reply dataflow/core/test/test_utils.py:146]: @samarth9008 This doesn't work, it ends up modifying message content.
oie_uaTvufbEpxzz

Comment (Jd8111997) [reply dataflow/core/test/test_utils.py:146]: @samarth9008 @gpsaggese Do you have any other suggestion ?

File: dev_scripts/client_setup/requirements.txt (hunk 15 above)
Comment (Jd8111997) [reply dev_scripts/client_setup/requirements.txt:10]: @gpsaggese Should I remove this ?

File: dataflow/core/test/test_utils.py (hunk 19 above)
Comment (sonaalKant) [reply dataflow/core/test/test_utils.py:146]: Try this. `self.assert_equal(actual, expected, fuzzy_match=True)`

Comment (Jd8111997) [reply dataflow/core/test/test_utils.py:146]: Thanks @sonaalKant, it worked!

File: dev_scripts/client_setup/requirements.txt (hunk 15 above)
Comment (Jd8111997) [reply dev_scripts/client_setup/requirements.txt:10]: @sonaalKant Shall I remove this or keep it ?

**Comment by samarth9008** [comment]
@Jd8111997 Once the changes are resolved, always request the review from the person who requested the changes.

File: dev_scripts/client_setup/requirements.txt (hunk 15 above)
Comment (samarth9008) [reply dev_scripts/client_setup/requirements.txt:10]: You can keep it.


@gpsaggese This is a usual dependency on sorrentum which is generating errors while building the env for new people. I would suggest we can keep it for now and can file follow-up to deal with this

@Jd8111997 Can you file an issue for this? No need to work on it. Just an issue.

**Comment by samarth9008** [comment]
Before merging I would suggest to create a PR in `cmamp` as we are touching a lot of core stuff

FYI @sonaalKant

File: dataflow/core/test/test_utils.py [hunk 21]
```diff
@@ -48,3 +49,147 @@ def test3(self) -> None:
...
+            dtfcorutil.convert_to_multiindex(df, asset_id_col)
+        actual_error_message = str(context.exception)
+        exp = r"""
+
+        ################################################################################
+        * Failed assertion *
+        1 <= 0
+        ################################################################################
+
+        """
```
Comment (samarth9008) [inline dataflow/core/test/test_utils.py:153]: Take a look how I have created this exp msg. Make sure it looks like this for other test as well.

File: dataflow/notebooks/build_dag_with_different_approaches.how_to_guide.py [hunk 22]
```diff
@@ -209,7 +209,7 @@ def resample_calculate_twap_vwap_and_returns(df, resampling_freq):
 
 # %% run_control={"marked": false}
 # Convert historical data to multiindex format.
-converted_data = dtfsys.source_nodes._convert_to_multiindex(
+converted_data = dtfcore.utils.convert_to_multiindex(
```
Comment (samarth9008) [inline dataflow/notebooks/build_dag_with_different_approaches.how_to_guide.py:212]: Does `dtfcore.convert_to_multiindex(..)` works? Check everywhere.

If it doesn't add import with utils at the top and use that to call the function directly. Same approach everywhere.

**Review by samarth9008 — COMMENTED** [review]
Recommendation and question

File: dev_scripts/client_setup/requirements.txt (hunk 15 above)
Comment (Jd8111997) [reply dev_scripts/client_setup/requirements.txt:10]: Absolutely @samarth9008.

File: dataflow/core/test/test_utils.py (hunk 21 above)
Comment (Jd8111997) [reply dataflow/core/test/test_utils.py:153]: Thanks, it did work!

File: dataflow/notebooks/build_dag_with_different_approaches.how_to_guide.py (hunk 22 above)
Comment (Jd8111997) [reply dataflow/notebooks/build_dag_with_different_approaches.how_to_guide.py:212]: @samarth9008 I am confused about this. So what are you saying is directly importing the whole package as ```dtfcore``` and calling ```convert_to_multiindex(..)``` everywhere ? Also what about your [previous suggestion](https://github.com/sorrentum/sorrentum/pull/712#discussion_r1518642013 ). Could you please elaborate on this ?

Comment (Jd8111997) [reply dataflow/notebooks/build_dag_with_different_approaches.how_to_guide.py:212]: @samarth9008 any update on this ?

Comment (samarth9008) [reply dataflow/notebooks/build_dag_with_different_approaches.how_to_guide.py:212]: > @samarth9008 I am confused about this. So what are you saying is directly importing the whole package as dtfcore and calling convert_to_multiindex(..) everywhere ?

Yes. If that doesn't work you can import the whole path to the file that contains the function.

Comment (samarth9008) [reply dataflow/notebooks/build_dag_with_different_approaches.how_to_guide.py:212]: Basically removing the utils in between.

Comment (Jd8111997) [reply dataflow/notebooks/build_dag_with_different_approaches.how_to_guide.py]: Done.
